File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/items/Item.java
package de.limbusdev.guardianmonsters.guardians.items;


import de.limbusdev.guardianmonsters.guardians.monsters.Monster;

/**
 * @author Georg Eckert on 17.02.17.
 */

public abstract class Item
{
    public enum Category
    {
        ALL, MEDICINE, EQUIPMENT, KEY,
    }

    private static int INSTANCE_COUNTER=0;

    public final int ID;
    private String name;
    private Category category;

    public Item(String name, Category category)
    {
        this.name = name;
        this.ID = INSTANCE_COUNTER;
        this.category = category;
        INSTANCE_COUNTER++;
    }

    public String getName() {
        return name;
    }

    public Category getCategory() {
        return category;
    }

    public abstract void apply(Monster m);

    public abstract boolean applicable(Monster m);

    @Override
    public boolean equals(Object obj)
    {
        if(obj instanceof Item) {
            return ((Item)obj).getName().equals(name);
        } else {
            return false;
        }
    }
}

File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/items/Item.kt
package de.limbusdev.guardianmonsters.guardians.items

/**
 * @author Georg Eckert on 17.02.17.
 */

abstract class Item(val name: String, val category: Category)
{
    val ID: Int

    init
    {
        this.ID = INSTANCE_COUNTER++
    }


    override fun equals(other: Any?): Boolean
    {
        return (other.hashCode() == hashCode())
    }

    override fun hashCode(): Int
    {
        var result = name.hashCode()
        result = 31 * result + category.hashCode()
        return result
    }

    companion object
    {
        private var INSTANCE_COUNTER = 0
    }

    // ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ Inner Classes
    enum class Category
    {
        ALL,
        MEDICINE,
        EQUIPMENT,
        KEY,
        CHAKRACRYSTAL
    }
}
